Library/ViewModels/DiscoveryPageViewModel.swift

```diff
@@ -142,36 +145,58 @@ public final class DiscoveryPageViewModel: DiscoveryPageViewModelType, Discovery
       .map { $0.1 }
 
     let requestFirstPageWith = Signal.merge(
-        firstPageParams,
-        (firstPageParams.takeWhen(environmentChanged)))
+      firstPageParams,
+      (firstPageParams.takeWhen(environmentChanged)),
+      firstPageParams.takeWhen(self.pulledToRefreshProperty.signal))
 
     let paginatedProjects: Signal<[Project], NoError>
     let pageCount: Signal<Int, NoError>
-    (paginatedProjects, self.projectsAreLoading, pageCount) = paginate(
+    let isLoading: Signal<Bool, NoError>
+    (paginatedProjects, isLoading, pageCount) = paginate(
       requestFirstPageWith: requestFirstPageWith,
       requestNextPageWhen: isCloseToBottom,
-      clearOnNewRequest: true,
+      clearOnNewRequest: false,
       skipRepeats: false,
       valuesFromEnvelope: { $0.projects },
       cursorFromEnvelope: { $0.urls.api.moreProjects },
       requestFromParams: { AppEnvironment.current.apiService.fetchDiscovery(params: $0) },
       requestFromCursor: { AppEnvironment.current.apiService.fetchDiscovery(paginationUrl: $0) },
       concater: { ($0 + $1).distincts() })
 
-      let projects = Signal.merge(
-        paginatedProjects,
-        self.selectedFilterProperty.signal.skipNil().skipRepeats().mapConst([])
+    let projects = Signal.merge(
+      paginatedProjects,
+      self.selectedFilterProperty.signal.skipNil().skipRepeats().mapConst([])
       )
       .skip { $0.isEmpty }
       .skipRepeats(==)
 
-      self.projectsLoaded =  self.selectedFilterProperty.signal
-        .takePairWhen(projects)
-        .map { ($1, $0) }
+    self.projectsLoaded =  self.selectedFilterProperty.signal
+      .takePairWhen(projects)
+      .map { ($1, $0) }
 
     self.asyncReloadData = self.projectsLoaded.take(first: 1).ignoreValues()
 
-    self.showEmptyState = Signal.combineLatest(paramsChanged, self.projectsAreLoading, paginatedProjects)
+    let isRefreshing = isLoading
+      .combineLatest(with: self.pulledToRefreshProperty.signal)
+      .map { $0.0 }
+      .skipRepeats()
+
+    let projectsLoadingNoRefresh = Signal.merge(
+      isLoading,
+      isLoading.takeWhen(isRefreshing).mapConst(false)
+    ).skipRepeats()
+
+    self.projectsAreLoadingAnimated = Signal.merge(
+      isRefreshing.map { ($0, true) },
+      projectsLoadingNoRefresh.map { ($0, false) },
+      self.viewWillAppearProperty.signal.take(first: 1).mapConst((true, false))
+    )
+    .skipRepeats(==)
```

Comment:
Can you explain this line and in particular `(==)`?

Reply:
In some cases, we need to pass a function that determines that values from the signal are equivalent. Without the `==`, the compiler will show the `missing argument` error.